// ethers/dist/ethers.umd.js
const __$G = (typeof globalThis !== 'undefined' ? globalThis: typeof window !== 'undefined' ? window: typeof global !== 'undefined' ? global: typeof self !== 'undefined' ? self: {});
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define(['exports'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ethers = {}));
})(this, (function (exports) { 'use strict';

    /* Do NOT modify this file; see /src.ts/_admin/update-version.ts */
    /**
     *  The current version of Ethers.
     */
    const version = "6.16.0";

    /**
     *  Property helper functions.
     *
     *  @_subsection api/utils:Properties  [about-properties]
     */
    function checkType(value, type, name) {
        const types = type.split("|").map(t => t.trim());
        for (let i = 0; i < types.length; i++) {
            switch (type) {
                case "any":
                    return;
                case "bigint":
                case "boolean":
                case "number":
                case "string":
                    if (typeof (value) === type) {
                        return;
                    }
            }
        }
        const error = new Error(`invalid value for type ${type}`);
        error.code = "INVALID_ARGUMENT";
        error.argument = `value.${name}`;
        error.value = value;
        throw error;
    }
    /**
     *  Resolves to a new object that is a copy of %%value%%, but with all
     *  values resolved.
     */
    async function resolveProperties(value) {
        const keys = Object.keys(value);
        const results = await Promise.all(keys.map((k) => Promise.resolve(value[k])));
        return results.reduce((accum, v, index) => {
            accum[keys[index]] = v;
            return accum;
        }, {});
    }
    /**
     *  Assigns the %%values%% to %%target%% as read-only values.
     *
     *  It %%types%% is specified, the values are checked.
     */
    function defineProperties(target, values, types) {
        for (let key in values) {
            let value = values[key];
            const type = (types ? types[key] : null);
            if (type) {
                checkType(value, type, key);
            }
            Object.defineProperty(target, key, { enumerable: true, value, writable: false });
        }
    }

    /**
     *  All errors in ethers include properties to ensure they are both
     *  human-readable (i.e. ``.message``) and machine-readable (i.e. ``.code``).
     *
     *  The [[isError]] function can be used to check the error ``code`` and
     *  provide a type guard for the properties present on that error interface.
     *
     *  @_section: api/utils/errors:Errors  [about-errors]
     */
    function stringify$1(value, seen) {
        if (value == null) {
            return "null";
        }
        if (seen == null) {
            seen = new Set();
        }
        if (typeof (value) === "object") {
            if (seen.has(value)) {
                return "[Circular]";
            }
            seen.add(value);
        }
        if (Array.isArray(value)) {
            return "[ " + (value.map((v) => stringify$1(v, seen))).join(", ") + " ]";
        }
        if (value instanceof Uint8Array) {
            const HEX = "0123456789abcdef";
            let result = "0x";
            for (let i = 0; i < value.length; i++) {
                result += HEX[value[i] >> 4];
                result += HEX[value[i] & 0xf];
            }
            return result;
        }
        if (typeof (value) === "object" && typeof (value.toJSON) === "function") {
            return stringify$1(value.toJSON(), seen);
        }
        switch (typeof (value)) {
            case "boolean":
            case "number":
            case "symbol":
                return value.toString();
            case "bigint":
                return BigInt(value).toString();
            case "string":
                return JSON.stringify(value);
            case "object": {
                const keys = Object.keys(value);
                keys.sort();
                return "{ " + keys.map((k) => `${stringify$1(k, seen)}: ${stringify$1(value[k], seen)}`).join(", ") + " }";
            }
        }
        return `[ COULD NOT SERIALIZE ]`;
    }
    /**
     *  Returns true if the %%error%% matches an error thrown by ethers
     *  that matches the error %%code%%.
     *
     *  In TypeScript environments, this can be used to check that %%error%%
     *  matches an EthersError type, which means the expected properties will
     *  be set.
     *
     *  @See [ErrorCodes](api:ErrorCode)
     *  @example
     *    try {
     *      // code....
     *    } catch (e) {
     *      if (isError(e, "CALL_EXCEPTION")) {
     *          // The Type Guard has validated this object
     *          console.log(e.data);
     *      }
     *    }
     */
    function isError(error, code) {
        return (error && error.code === code);
    }
    /**
     *  Returns true if %%error%% is a [[CallExceptionError].
     */
    function isCallException(error) {
        return isError(error, "CALL_EXCEPTION");
    }
    /**
     *  Returns a new Error configured to the format ethers emits errors, with
     *  the %%message%%, [[api:ErrorCode]] %%code%% and additional properties
     *  for the corresponding EthersError.
     *
     *  Each error in ethers includes the version of ethers, a
     *  machine-readable [[ErrorCode]], and depending on %%code%%, additional
     *  required properties. The error message will also include the %%message%%,
     *  ethers version, %%code%% and all additional properties, serialized.
     */
    function makeError(message, code, info) {
        let shortMessage = message;
        {
            const details = [];
            if (info) {
                if ("message" in info || "code" in info || "name" in info) {
                    throw new Error(`value will overwrite populated values: ${stringify$1(info)}`);
                }
                for (const key in info) {
                    if (key === "shortMessage") {
                        continue;
                    }
                    const value = (info[key]);
                    //                try {
                    details.push(key + "=" + stringify$1(value));
                    //                } catch (error: any) {
                    //                console.log("MMM", error.message);
                    //                    details.push(key + "=[could not serialize object]");
                    //                }
                }
            }
            details.push(`code=${code}`);
            details.push(`version=${version}`);
            if (details.length) {
                message += " (" + details.join(", ") + ")";
            }
        }
        let error;
        switch (code) {
            case "INVALID_ARGUMENT":
                error = new TypeError(message);
                break;
            case "NUMERIC_FAULT":
            case "BUFFER_OVERRUN":
                error = new RangeError(message);
                break;
            default:
                error = new Error(message);
        }
        defineProperties(error, { code });
        if (info) {
            Object.assign(error, info);
        }
        if (error.shortMessage == null) {
            defineProperties(error, { shortMessage });
        }
        return error;
    }
    /**
     *  Throws an EthersError with %%message%%, %%code%% and additional error
     *  %%info%% when %%check%% is falsish..
     *
     *  @see [[api:makeError]]
     */
    function assert(check, message, code, info) {
        if (!check) {
            throw makeError(message, code, info);
        }
    }
    /**
     *  A simple helper to simply ensuring provided arguments match expected
     *  constraints, throwing if not.
     *
     *  In TypeScript environments, the %%check%% has been asserted true, so
     *  any further code does not need additional compile-time checks.
     */
    function assertArgument(check, message, name, value) {
        assert(check, message, "INVALID_ARGUMENT", { argument: name, value: value });
    }
    function assertArgumentCount(count, expectedCount, message) {
        if (message == null) {
            message = "";
        }
        if (message) {
            message = ": " + message;
        }
        assert(count >= expectedCount, "missing argument" + message, "MISSING_ARGUMENT", {
            count: count,
            expectedCount: expectedCount
        });
        assert(count <= expectedCount, "too many arguments" + message, "UNEXPECTED_ARGUMENT", {
            count: count,
            expectedCount: expectedCount
        });
    }
    const _normalizeForms = ["NFD", "NFC", "NFKD", "NFKC"].reduce((accum, form) => {
        try {
            // General test for normalize
            /* c8 ignore start */
            if ("test".normalize(form) !== "test") {
                throw new Error("bad");
            }
            ;
            /* c8 ignore stop */
            if (form === "NFD") {
                const check = String.fromCharCode(0xe9).normalize("NFD");
                const expected = String.fromCharCode(0x65, 0x0301);
                /* c8 ignore start */
                if (check !== expected) {
                    throw new Error("broken");
                }
                /* c8 ignore stop */
            }
            accum.push(form);
        }
        catch (error) { }
        return accum;
    }, []);
    /**
     *  Throws if the normalization %%form%% is not supported.
     */
    function assertNormalize(form) {
        assert(_normalizeForms.indexOf(form) >= 0, "platform missing String.prototype.normalize", "UNSUPPORTED_OPERATION", {
            operation: "String.prototype.normalize", info: { form }
        });
    }
    /**
     *  Many classes use file-scoped values to guard the constructor,
     *  making it effectively private. This facilitates that pattern
     *  by ensuring the %%givenGaurd%% matches the file-scoped %%guard%%,
     *  throwing if not, indicating the %%className%% if provided.
     */
    function assertPrivate(givenGuard, guard, className) {
        if (className == null) {
            className = "";
        }
        if (givenGuard !== guard) {
            let method = className, operation = "new";
            if (className) {
                method += ".";
                operation += " " + className;
            }
            assert(false, `private constructor; use ${method}from* methods`, "UNSUPPORTED_OPERATION", {
                operation
            });
        }
    }

    /**
     *  Some data helpers.
     *
     *
     *  @_subsection api/utils:Data Helpers  [about-data]
     */
    function _getBytes(value, name, copy) {
        if (value instanceof Uint8Array) {
            if (copy) {
                return new Uint8Array(value);
            }
            return value;
        }
        if (typeof (value) === "string" && (value.length % 2) === 0 &&
            value.match(/^0x[0-9a-f]*$/i)) {
            const result = new Uint8Array((value.length - 2) / 2);
            let offset = 2;
            for (let i = 0; i < result.length; i++) {
                result[i] = parseInt(value.substring(offset, offset + 2), 16);
                offset += 2;
            }
            return result;
        }
        assertArgument(false, "invalid BytesLike value", name || "value", value);
    }
    /**
     *  Get a typed Uint8Array for %%value%%. If already a Uint8Array
     *  the original %%value%% is returned; if a copy is required use
     *  [[getBytesCopy]].
     *
     *  @see: getBytesCopy
     */
    function getBytes(value, name) {
        return _getBytes(value, name, false);
    }
    /**
     *  Get a typed Uint8Array for %%value%%, creating a copy if necessary
     *  to prevent any modifications of the returned value from being
     *  reflected elsewhere.
     *
     *  @see: getBytes
     */
    function getBytesCopy(value, name) {
        return _getBytes(value, name, true);
    }
    /**
     *  Returns true if %%value%% is a valid [[HexString]].
     *
     *  If %%length%% is ``true`` or a //number//, it also checks that
     *  %%value%% is a valid [[DataHexString]] of %%length%% (if a //number//)
     *  bytes of data (e.g. ``0x1234`` is 2 bytes).
     */
    function isHexString(value, length) {
        if (typeof (value) !== "string" || !value.match(/^0x[0-9A-Fa-f]*$/)) {
            return false;
        }
        if (typeof (length) === "number" && value.length !== 2 + 2 * length) {
            return false;
        }
        if (length === true && (value.length % 2) !== 0) {
            return false;
        }
        return true;
    }
    /**
     *  Returns true if %%value%% is a valid representation of arbitrary
     *  data (i.e. a valid [[DataHexString]] or a Uint8Array).
     */
    function isBytesLike(value) {
        return (isHexString(value, true) || (value instanceof Uint8Array));
    }
    const HexCharacters = "0123456789abcdef";
    /**
     *  Returns a [[DataHexString]] representation of %%data%%.
     */
    function hexlify(data) {
        const bytes = getBytes(data);
        let result = "0x";
        for (let i = 0; i < bytes.length; i++) {
            const v = bytes[i];
            result += HexCharacters[(v & 0xf0) >> 4] + HexCharacters[v & 0x0f];
        }
        return result;
    }
    /**
     *  Returns a [[DataHexString]] by concatenating all values
     *  within %%data%%.
     */
    function concat(datas) {
        return "0x" + datas.map((d) => hexlify(d).substring(2)).join("");
    }
    /**
     *  Returns the length of %%data%%, in bytes.
     */
    function dataLength(data) {
        if (isHexString(data, true)) {
            return (data.length - 2) / 2;
        }
        return getBytes(data).length;
    }
    /**
     *  Returns a [[DataHexString]] by slicing %%data%% from the %%start%%
     *  offset to the %%end%% offset.
     *
     *  By default %%start%% is 0 and %%end%% is the length of %%data%%.
     */
    function dataSlice(data, start, end) {
        const bytes = getBytes(data);
        if (end != null && end > bytes.length) {
            assert(false, "cannot slice beyond data bounds", "BUFFER_OVERRUN", {
                buffer: bytes, length: bytes.length, offset: end
            });
        }
        return hexlify(bytes.slice((start == null) ? 0 : start, (end == null) ? bytes.length : end));
    }
    /**
     *  Return the [[DataHexString]] result by stripping all **leading**
     ** zero bytes from %%data%%.
     */
    function stripZerosLeft(data) {
        let bytes = hexlify(data).substring(2);
        while (bytes.startsWith("00")) {
            bytes = bytes.substring(2);
        }
        return "0x" + bytes;
    }
    function zeroPad(data, length, left) {
        const bytes = getBytes(data);
        assert(length >= bytes.length, "padding exceeds data length", "BUFFER_OVERRUN", {
            buffer: new Uint8Array(bytes),
            length: length,
            offset: length + 1
        });
        const result = new Uint8Array(length);
        result.fill(0);
        if (left) {
            result.set(bytes, length - bytes.length);
        }
        else {
            result.set(bytes, 0);
        }
        return hexlify(result);
    }
    /**
     *  Return the [[DataHexString]] of %%data%% padded on the **left**
     *  to %%length%% bytes.
     *
     *  If %%data%% already exceeds %%length%%, a [[BufferOverrunError]] is
     *  thrown.
     *
     *  This pads data the same as **values** are in Solidity
     *  (e.g. ``uint128``).
     */
    function zeroPadValue(data, length) {
        return zeroPad(data, length, true);
    }
    /**
     *  Return the [[DataHexString]] of %%data%% padded on the **right**
     *  to %%length%% bytes.
     *
     *  If %%data%% already exceeds %%length%%, a [[BufferOverrunError]] is
     *  thrown.
     *
     *  This pads data the same as **bytes** are in Solidity
     *  (e.g. ``bytes16``).
     */
    function zeroPadBytes(data, length) {
        return zeroPad(data, length, false);
    }

    /**
     *  Some mathematic operations.
     *
     *  @_subsection: api/utils:Math Helpers  [about-maths]
     */
    const BN_0$a = BigInt(0);
    const BN_1$5 = BigInt(1);
    //const BN_Max256 = (BN_1 << BigInt(256)) - BN_1;
    // IEEE 754 support 53-bits of mantissa
    const maxValue = 0x1fffffffffffff;
    /**
     *  Convert %%value%% from a twos-compliment representation of %%width%%
     *  bits to its value.
     *
     *  If the highest bit is ``1``, the result will be negative.
     */
    function fromTwos(_value, _width) {
        const value = getUint(_value, "value");
        const width = BigInt(getNumber(_width, "width"));
        assert((value >> width) === BN_0$a, "overflow", "NUMERIC_FAULT", {
            operation: "fromTwos", fault: "overflow", value: _value
        });
        // Top bit set; treat as a negative value
        if (value >> (width - BN_1$5)) {
            const mask = (BN_1$5 << width) - BN_1$5;
            return -(((~value) & mask) + BN_1$5);
        }
        return value;
    }
    /**
     *  Convert %%value%% to a twos-compliment representation of
     *  %%width%% bits.
     *
     *  The result will always be positive.
     */
    function toTwos(_value, _width) {
        let value = getBigInt(_value, "value");
        const width = BigInt(getNumber(_width, "width"));
        const limit = (BN_1$5 << (width - BN_1$5));
        if (value < BN_0$a) {
            value = -value;
            assert(value <= limit, "too low", "NUMERIC_FAULT", {
                operation: "toTwos", fault: "overflow", value: _value
            });
            const mask = (BN_1$5 << width) - BN_1$5;
            return ((~value) & mask) + BN_1$5;
        }
        else {
            assert(value < limit, "too high", "NUMERIC_FAULT", {
                operation: "toTwos", fault: "overflow", value: _value
            });
        }
        return value;
    }
    /**
     *  Mask %%value%% with a bitmask of %%bits%% ones.
     */
    function mask(_value, _bits) {
        const value = getUint(_value, "value");
        const bits = BigInt(getNumber(_bits, "bits"));
        return value & ((BN_1$5 << bits) - BN_1$5);
    }
    /**
     *  Gets a BigInt from %%value%%. If it is an invalid value for
     *  a BigInt, then an ArgumentError will be thrown for %%name%%.
     */
    function getBigInt(value, name) {
        switch (typeof (value)) {
            case "bigint": return value;
            case "number":
                assertArgument(Number.isInteger(value), "underflow", name || "value", value);
                assertArgument(value >= -maxValue && value <= maxValue, "overflow", name || "value", value);
                return BigInt(value);
            case "string":
                try {
                    if (value === "") {
                        throw new Error("empty string");
                    }
                    if (value[0] === "-" && value[1] !== "-") {
                        return -BigInt(value.substring(1));
                    }
                    return BigInt(value);
                }
                catch (e) {
                    assertArgument(false, `invalid BigNumberish string: ${e.message}`, name || "value", value);
                }
        }
        assertArgument(false, "invalid BigNumberish value", name || "value", value);
    }
    /**
     *  Returns %%value%% as a bigint, validating it is valid as a bigint
     *  value and that it is positive.
     */
    function getUint(value, name) {
        const result = getBigInt(value, name);
        assert(result >= BN_0$a, "unsigned value cannot be negative", "NUMERIC_FAULT", {
            fault: "overflow", operation: "getUint", value
        });
        return result;
    }
    const Nibbles$1 = "0123456789abcdef";
    /*
     * Converts %%value%% to a BigInt. If %%value%% is a Uint8Array, it
     * is treated as Big Endian data.
     */
    function toBigInt(value) {
        if (value instanceof Uint8Array) {
            let result = "0x0";
            for (const v of value) {
